File: nodejs-tools/nodejs-compiler-tools/formatter-xml/src/format_xml_expression.test.ts
import { expect, test } from "vitest";
import {
  XmlCommentNode,
  XmlPrologNode,
  XmlTagNode,
} from "@vighnesh153/parser-xml";
import { formatXmlExpression } from "./format_xml_expression.ts";
import { parseProgram } from "./test_utils.ts";

test("should format xml prolog node", () => {
  const [parser, program] = parseProgram(
    `<? xml  encoding =   "utf-8"  version =  "2.0"   ?>`,
  );

  expect(parser.errors.length).toBe(0);
  expect(program.statements.length).toBe(1);
  expect(
    formatXmlExpression({
      expression: program.statements[0] as XmlPrologNode,
      indentation: 4,
      indentationLevel: 0,
      sortAttributes: true,
    }),
  ).toMatchInlineSnapshot(`"<?xml encoding="utf-8" version="2.0"?>"`);
});

test("should format xml tag node", () => {
  const [parser, program] = parseProgram(
    `<manifest   simpleProperty = "some random value" deeply :  nested:property = "20" >
  </ manifest>
    `,
  );

  expect(parser.errors.length).toBe(0);
  expect(program.statements.length).toBe(1);
  expect(
    formatXmlExpression({
      expression: program.statements[0] as XmlTagNode,
      indentation: 4,
      indentationLevel: 0,
      sortAttributes: true,
    }),
  ).toMatchInlineSnapshot(`
    "<manifest
        simpleProperty="some random value"
        deeply:nested:property="20" />"
  `);
});

test("should format comment node", () => {
  const [parser, program] = parseProgram(
    `<!-- <uses-permission android:name="com.android.providers.tv.permission.WRITE_EPG_DATA" /> -->`,
  );

  expect(parser.errors.length).toBe(0);
  expect(program.statements.length).toBe(1);
  expect(
    formatXmlExpression({
      expression: program.statements[0] as XmlCommentNode,
      indentation: 4,
      indentationLevel: 0,
      sortAttributes: true,
    }),
  ).toMatchInlineSnapshot(
    `"<!-- <uses-permission android:name="com.android.providers.tv.permission.WRITE_EPG_DATA" /> -->"`,
  );
});

test("should format text node", () => {
  const [parser, program] = parseProgram(
    `< pokemon  >  Pikachu    < /  pokemon >`,
  );

  expect(parser.errors.length).toBe(0);
  expect(program.statements.length).toBe(1);
  expect(
    formatXmlExpression({
      expression: (program.statements[0] as XmlTagNode).children[0],
      indentation: 4,
      indentationLevel: 0,
      sortAttributes: true,
    }),
  ).toMatchInlineSnapshot(`"Pikachu"`);
});

File: jvm-tools/XmlFormatter/src/test/kotlin/dev/vighnesh153/xml/formatter/fmt/FormatXmlExpressionTest.kt
package dev.vighnesh153.xml.formatter.fmt

import dev.vighnesh153.xml.formatter.ast.XmlProgram
import dev.vighnesh153.xml.formatter.ast.XmlTagNode
import dev.vighnesh153.xml.formatter.lexer.LexerInputReader
import dev.vighnesh153.xml.formatter.lexer.XmlLexer
import dev.vighnesh153.xml.formatter.parser.XmlParser
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class FormatXmlExpressionTest {
    private lateinit var lexer: XmlLexer
    private lateinit var parser: XmlParser
    private lateinit var program: XmlProgram

    @Test
    fun formatsXmlPrologNode() {
        setXml("""<? xml  encoding =   "utf-8"  version =  "2.0"   ?>""")

        assertEquals(0, parser.errors.size)
        assertEquals(1, program.statements.size)
        assertEquals(
            """<?xml encoding="utf-8" version="2.0"?>""",
            program.statements.first().format(
                indentationLevel = 0,
                indentation = 4,
                sortAttributes = true
            )
        )
    }

    @Test
    fun formatsXmlTagNode() {
        setXml(
            """
           <manifest   simpleProperty = "some random value" deeply :  nested:property = "20" >
                    </ manifest> 
        """
        )

        assertEquals(0, parser.errors.size)
        assertEquals(1, program.statements.size)
        assertEquals(
            """
                <manifest
                    simpleProperty="some random value"
                    deeply:nested:property="20" />
            """.trimIndent(),
            program.statements.first().format(
                indentationLevel = 0,
                indentation = 4,
                sortAttributes = true
            )
        )
    }

    @Test
    fun formatsXmlCommentNode() {
        setXml(
            """
           <!--
            
            
            
<uses-permission android:name="com.android.providers.tv.permission.WRITE_EPG_DATA" /> 



--> 
        """
        )

        assertEquals(0, parser.errors.size)
        assertEquals(1, program.statements.size)
        assertEquals(
            """
                <!-- <uses-permission android:name="com.android.providers.tv.permission.WRITE_EPG_DATA" /> -->
            """.trimIndent(),
            program.statements.first().format(
                indentationLevel = 0,
                indentation = 4,
                sortAttributes = true
            )
        )
    }

    @Test
    fun formatsXmlTextNode() {
        setXml("< pokemon  >  Pikachu    < /  pokemon >")

        assertEquals(0, parser.errors.size)
        assertEquals(1, program.statements.size)
        assertEquals(
            "Pikachu",
            (program.statements.first() as XmlTagNode).children.first().format(
                indentationLevel = 0,
                indentation = 4,
                sortAttributes = true
            )
        )
    }

    private fun setXml(input: String) {
        lexer = XmlLexer(LexerInputReader(input))
        parser = XmlParser(lexer)
        program = parser.parseProgram()
    }
}